File: Data Structures & Algorithms/combination-target-sum/submission-25.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) { 
        const res = [] 
        const part = [] 
        this.dfs(0, nums, target, part ,res) 
        return res
    } 

    dfs(start, nums, target, part, res) {
        if (target === 0) {
            res.push([...part]) 
            return
        }  

        if (target < 0) {
            return
        }

        for (let i = start; i < nums.length; i ++) {
            part.push(nums[i]) 
            target -= nums[i] 
            this.dfs(i, nums, target, part, res) 
            part.pop() 
            target += nums[i] 
        }
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-25.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        
        def dfs(i, prev): 
            if i == len(nums): 
                return 0 

            res = dfs(i + 1, prev) 

            if prev == float('inf') or nums[i] > prev: 
                res = max(res, 1 + dfs(i + 1, nums[i])) 

            return res 

        return dfs(0, float('inf'))